cpp/src/strings/capitalize.cu

```diff
@@ -0,0 +1,356 @@
+      d_case_table_(d_case_table)
+    {  
+    }
+
+    __host__ __device__
+    column_device_view const get_column(void) const
+    {
+      return d_column_;
+    }
+
+    __host__ __device__
+    character_flags_table_type const* get_flags(void) const
+    {
+      return d_flags_;
+    }
+
+    __host__ __device__
+    character_cases_table_type const* get_case_table(void) const
+    {
+      return d_case_table_;
+    }
+
+    __device__
+    virtual int32_t operator()(size_type idx) const = 0;
+    
+  private:
+    column_device_view const d_column_;
+    character_flags_table_type const* d_flags_;
+    character_cases_table_type const* d_case_table_;
+  };
+
+  
+  //base class for executing string modification:
+  //
+  struct execute_base
+  {
+    execute_base(column_device_view const d_column,
+                 character_flags_table_type const* d_flags,
+                 character_cases_table_type const* d_case_table,
+                 int32_t const* d_offsets,
+                 char* d_chars):
+      d_column_(d_column),
+      d_flags_(d_flags),
+      d_case_table_(d_case_table),
+      d_offsets_(d_offsets),
+      d_chars_(d_chars)
+    {
+    }
+
+    __host__ __device__
+    column_device_view const get_column(void) const
+    {
+      return d_column_;
+    }
+
+    __host__ __device__
+    character_flags_table_type const* get_flags(void) const
+    {
+      return d_flags_;
+    }
+
+    __host__ __device__
+    character_cases_table_type const* get_case_table(void) const
+    {
+      return d_case_table_;
+    }
+
+    __host__ __device__
+    int32_t const* get_offsets(void) const
+    {
+      return d_offsets_;
+    }
+
+    __host__ __device__
+    char* get_chars(void)
+    {
+      return d_chars_;
+    }
+
+    __device__
+    virtual int32_t operator()(size_type idx) = 0;
+  private:
+    column_device_view const d_column_;
+    character_flags_table_type const* d_flags_;
+    character_cases_table_type const* d_case_table_;
+    int32_t const* d_offsets_;
+    char* d_chars_;
+  };
+
+
+  //functor for probing string capitalization
+  //requirements:
+  //
+  struct probe_capitalize: probe_base
+  {
+    probe_capitalize(column_device_view const d_column,
+                     character_flags_table_type const* d_flags,
+                     character_cases_table_type const* d_case_table):
+      probe_base(d_column, d_flags, d_case_table)
+    {  
+    }
+    
+     __device__
+     int32_t operator()(size_type idx) const override {
+       if( get_column().is_null(idx) )
+         return 0; // null string
+      
+       string_view d_str = get_column().template element<string_view>(idx);
+       int32_t bytes = 0;
+      
+       for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+         auto the_chr = *itr;
+         uint32_t code_point = detail::utf8_to_codepoint(the_chr);
+         detail::character_flags_table_type flag = code_point <= 0x00FFFF ? get_flags()[code_point] : 0;
```

Comment:
Perhaps this line could be made into a base-class method. The accessor method would no longer be required.
Maybe something like:
```
using char_info = thrust::pair<uint32_t,detail::character_flags_table_type>;
    __device__ char_info get_char_info(char_utf8 chr)
    {
        uint32_t code_point = detail::utf8_to_codepoint(chr);
        detail::character_flags_table_type flag = code_point <= 0x00FFFF ? d_flags[code_point] : 0;
        return char_info{code_point,flag};
    }
```

Reply:
Good idea. Done. BTW: accessor functions are also zero-cost because they are inlined and optimized away by the compiler. Nevertheless, factoring this code out is a good idea.

Reply:
Okay, replaced that everywhere.